pkg/kv/kvserver/rangefeed/unbuffered_registration.go

```diff
@@ -0,0 +1,497 @@
+// yet.
+func (ubr *unbufferedRegistration) waitForCaughtUp(ctx context.Context) error {
+	opts := retry.Options{
+		InitialBackoff: 5 * time.Millisecond,
+		Multiplier:     2,
+		MaxBackoff:     10 * time.Second,
+		MaxRetries:     50,
+	}
+	for re := retry.StartWithCtx(ctx, opts); re.Next(); {
+		ubr.mu.Lock()
+		caughtUp := len(ubr.mu.catchUpBuf) == 0 && ubr.mu.caughtUp
+		ubr.mu.Unlock()
+		if caughtUp {
+			return nil
+		}
+	}
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+	return errors.Errorf("unbufferedRegistration %v failed to empty in time", ubr.Range())
+}
+
+func (ubr *unbufferedRegistration) setDisconnectedIfNot() {
+	ubr.mu.Lock()
+	defer ubr.mu.Unlock()
+	ubr.setDisconnectedIfNotWithLock()
+}
+
+// setDisconnectedIfNotWithLock sets disconnected to true if it is not already
+// set. It handles registration level cleanup. It is called from BufferedSender
+// to disconnect registrations. It is no-op if ubr.disconnect already happens.
+//
+// Note it is called while holding ubr.mu.
+func (ubr *unbufferedRegistration) setDisconnectedIfNotWithLock() (alreadyDisconnected bool) {
+	// TODO(wenyihu6 during code reviews or later on): think about if we should
+	// just drain catchUpBuf here during reviews. We never publish anything in
+	// catch-up buf if disconnected. But this might take a long time and you are
+	// on a hot path.
+	if ubr.mu.disconnected {
+		return true
+	}
+	if ubr.mu.catchUpIter != nil {
+		// catch-up scan hasn't started yet.
+		ubr.mu.catchUpIter.Close()
+		ubr.mu.catchUpIter = nil
+	}
+	if ubr.mu.catchUpScanCancelFn != nil {
+		ubr.mu.catchUpScanCancelFn()
+	}
+	ubr.mu.disconnected = true
+	return false
+}
+
+// discardCatchUpBufferWithLock drains catchUpBuf and release all memory held by
+// the events in catch-up buffer without publishing them. It is safe to assume
+// that catch-up buffer is empty and nil after call.
+func (ubr *unbufferedRegistration) discardCatchUpBufferWithLock(ctx context.Context) {
+	if ubr.mu.catchUpBuf == nil {
+		// Already drained.
+		return
+	}
+	func() {
+		for {
+			select {
+			case e := <-ubr.mu.catchUpBuf:
+				// TODO(wenyihu6): check if release with context.Background is fine
+				// during reviews
+				e.alloc.Release(ctx)
+				putPooledSharedEvent(e)
+			default:
+				// Done.
+				return
+			}
+		}
+	}()
+
+	ubr.mu.catchUpBuf = nil
+	ubr.mu.caughtUp = true
+}
+
+// publishCatchUpBuffer drains catchUpBuf and release all memory held by
+// the events in catch-up buffer while publishing them.
+//
+// Caller is responsible for draining it again if error is returned. If no
+// error, it is safe to assume that catch-up buffer is empty and nil after this
+// call.
+func (ubr *unbufferedRegistration) publishCatchUpBuffer(ctx context.Context) error {
+	// TODO(wenyihu6): check if we can just drain without holding the lock first
+	// during reviews We shouldn't be reading from the buffer at the same time
```

Comment:
I’m not sure if this is a valid concern, but I was worried about publishCatchUpBuffer being called concurrently with discardCatchUpBuf.  Both functions use channels, but they are both reading from the same channel. My concern was whether discardCatchUpBuf could drop some events while publishCatchUpBuffer is still running and sending more events, leading to dropped events. With the current setup, this shouldn’t happen since both functions are in the same goroutine. However, it might become an issue if we move discardCatchUpBuf to drainAllocations, which can be called from other goroutines.